### ISSUE TITLE
Displaying image message from network

### ISSUE THREAD

**Issue by vpadmnabh** [badoo/Chatto#436]
i have gone through previous issues regarding same problem. But none of them clearly provides a way to do this. I have existing Model where I am adding Chatto for UI so I am able to create a message with placeholder but unable to figure out the best way to change the same after download.

Even if i update the model message and perform chatDataSourceDidUpdate(self) it's not reflected.

My another question is what's the best way to access MessageViewModel at run time and is it a good idea.

**Comment by AntonPalich** [comment]
Each message is controlled by the corresponding presenter. You have access to a message view model from a presenter and you can modify a view model at any time here.

I suggest you to subclass `PhotoMessagePresenter` and override `cellWillBeShown` and `cellWasHidden` functions.

You can put a placeholder to a view model in `cellWillBeShown` function and start downloading here and you can cancel download (if you want) in `cellWasHidden` function. Your model should have an image URL. A small example:

```swift
    public override func cellWillBeShown() {
        super.cellWillBeShown()
        self.messageViewModel.image.value = // your placeholder
        self.downloadImage(from: self.messageModel.imageUrl) { [weak self] (image) in
            self?.messageViewModel.image.value = image
        }
    }
```

It's not a production ready code, but just a direction to move. Just remember, presenter should contain a business logic for a single message. You can pass all required dependencies (service that will download images, for example) in a init method of a presenter. Presenter is created by a presenter builder that you're registering for a spricific type of messages.

**Comment by vpadmnabh** [comment]
Thanks 
I'll give it a try.
Is there a way i can keep downloading logic separate so that my previous work is not obsolete. Or this is the only way we can do for downloading images. It's not necessary for me to show loading indicator right now.

**Comment by AntonPalich** [comment]
Of course, you can keep it separate. You just need to use it within presenter.

In Badoo, we use a separate service to download and cache images and we pass this service as a dependency into a presenter. When a photo message is about to appear (`cellWillBeShown` is called) we ask the service to download an image from URL taken from a message. When image is downloaded we assign this image into a view model and then cell is updated
